Add: 소셜 로그인 기능 구현

kyeongraekim opened the pull request:
## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
-  소셜 로그인 기능 구현
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
-  유저 이메일을 중심으로한 소셜 로그인 기능 구현
- MockResponse를 이용한 유닛테스트 성공

## 기타 질문 및 특이 사항
- 로그인은 일반로그인과 소셜로그인을 구분지었지만, 회원가입은 소셜 회원가입을 일반 회원가입에 통합시켰습니다.|
  소셜 회원가입도 따로 분리시키는게 나을지 궁금합니다.

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.

<img width="563" alt="소셜로그인 유닛테스트 성공" src="https://user-images.githubusercontent.com/73090474/120098732-911d9680-c172-11eb-8634-dde349433983.png">

Kang-kyunghun commented on users/views.py:
```diff
@@ -1,3 +1,49 @@
+class KakaoLogInView(View):
+    def get(self, request):
+        try:
+            access_token          = request.headers.get('Authorization', None)
+            headers               = {"Authorization" : f"Bearer {access_token}"}
+            kakao_login_data      = requests.get("https://kapi.kakao.com/v2/user/me", headers=headers).json()
+            kakao_account         = kakao_login_data["kakao_account"]
+            email                 = kakao_account.get("email")  
```
가장 긴 변수를 기준으로 정렬하면 좋을 것 같아요
```suggestion
            access_token      = request.headers.get('Authorization', None)
            headers           = {"Authorization" : f"Bearer {access_token}"}
            kakao_login_data  = requests.get("https://kapi.kakao.com/v2/user/me", headers=headers).json()
            kakao_account     = kakao_login_data["kakao_account"]
            email             = kakao_account.get("email")  
```

kyeongraekim replied:
네 수정 반영하겠습니다.

Kang-kyunghun commented on users/views.py:
```diff
@@ -1,3 +1,49 @@
+            if not kakao_login_data:   
+                return JsonResponse({'message' : 'INVALID_TOKEN'}, status = 401)
+
+            if kakao_login_data.get('code') == -401:  
```
코드가 -401이 있나요?(궁금)

kyeongraekim replied:
넵 카카오 디벨로퍼스 사이트를 확인해보니 카카오 자체 에러메세지 인것 같고,
아래의 상황에서 사용하는 것이라고 설명이 되어있습니다.

"유효하지 않은 앱키나 액세스 토큰으로 요청한 경우
토큰 값이 잘못되었거나 만료되어 유효하지 않은 경우로 토큰 갱신 필요"

Kang-kyunghun commented on users/views.py:
```diff
@@ -1,3 +1,49 @@
+
+            if not User.objects.filter(email=email).exists() and email != None:
+                return JsonResponse({'message' : 'INVALID_USER'}, status=401)
+
```
소셜로그인으로 로그인 하는데, 그 조건이 이미 소셜 로그인으로 로그인 할 이메일이 가입되어 있어야 되는 조건으로 기획 한건가요?

kyeongraekim replied:
네 맞습니다. 저희 조 기획상 일반/소셜 회원가입 시에 유저의 관련 정보가 모두 은행의 집 db에 들어오도록 되어있습니다.
따라서 카카오로 로그인을 하여도 저희 db에 해당 유저의 이메일이 기존재할 것이라 판단하여 작성한 코드입니다.

위 방법 말고도, 카카오 로그인 뷰에서 해당 유저가 없으면 가입을 바로 하게끔 하는 방법이 차선책으로 준비가 되어있어서,
현재 진행중인 방안이 여러 변수로 인해서 구현이 힘들게 된다면 get or create 나 if not ~~ exists()문을 써서 
카카오로그인 뷰에서 회원가입을 가능케 하겠습니다.

choidaehwan committed:
Fix: users앱 views.py 리뷰반영 수정
 - 리베이스시 다른 앱에 저장된 불필요한 코드, 파일 삭제
 - 리뷰반영 users앱 views.py 수정
 - 유닛테스트 확인

Kang-kyunghun commented on users/views.py:
```diff
@@ -1,3 +1,50 @@
+
+            kakao_id         = str(kakao_login_data["id"])
+            kakao_account    = kakao_login_data["kakao_account"]
+            email            = kakao_account.get("email", None)
```
꼭 필요한 값이므로 get 대신에 []으로 바꿔주세요.

Kang-kyunghun commented on users/tests.py:
```diff
@@ -1,3 +1,82 @@
+        header              = {"HTTP_Authorization" : "access_token"}
+        response            = client.get('/users/log-in/kakao', content_type='application/json', **header)
+        
+        self.assertEqual(response.status_code, 200)
```
코드만 아니고 출력 결과도 포함시켜 주세요.

Kang-kyunghun commented on posts/views.py:
```diff
@@ -1,3 +1,44 @@
+            user    = request.user
+            img_url = profile_image                  # 프로필 사진
+            date    = created_at                     # 
+            like    = 
+
+
```
1. 공백 제거
2. 코드가 완성이 안 되어 있네요.

kyeongraekim replied:
네 리베이스 도중 착오가 생겨서 개인 서브 리포지토리에 올릴 내용을 잘못 올린 것 같습니다.
수정 후 재업로드하겠습니다

Kang-kyunghun commented:
소셜로그인 브렌치인데 코멘트 코드까지 같이 있습니다. 브렌치 나누어서 따로 push 해주세요.

kyeongraekim committed:
리베이스 꼬임 해결용 임시저장

kyeongraekim committed:
리베이스 꼬임 해결용 전체 풀 받기
Merge branch 'main' of https://github.com/wecode-bootcamp-korea/20-2nd-BankHouse-backend

kyeongraekim committed:
Fix: 불필요한 파일 코드 삭제

kyeongraekim committed:
Add: git pull 내려받기
